java/src/jni/h5util.c

```diff
@@ -4079,6 +4081,521 @@ translate_rbuf(JNIEnv *env, jobjectArray ret_buf, jlong mem_type_id, H5T_class_t
+            }
+            break;
+        } /* H5T_INTEGER */
+        case H5T_FLOAT: {
+            /* Convert each element to a list */
+            jobject jobj = NULL;
+            switch (typeSize) {
+                case sizeof(jfloat): {
+                    jfloat floatValue;
+                    HDmemcpy(((char *)&floatValue), ((char *)raw_buf), typeSize);
+
+                    if (NULL == (jobj = ENVPTR->CallStaticObjectMethod(ENVONLY, cFloat, floatValueMid,
+                                                                       (double)floatValue)))
+                        CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+                    break;
+                }
+                case sizeof(jdouble): {
+                    jdouble doubleValue;
+                    HDmemcpy(((char *)&doubleValue), ((char *)raw_buf), typeSize);
+
+                    if (NULL == (jobj = ENVPTR->CallStaticObjectMethod(ENVONLY, cDouble, doubleValueMid,
+                                                                       doubleValue)))
+                        CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+                    break;
+                }
+            }
+            if (jobj) {
+                ENVPTR->CallBooleanMethod(ENVONLY, ret_buf, arrAddMethod, jobj);
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
+                ENVPTR->DeleteLocalRef(ENVONLY, jobj);
+            }
+            break;
+        } /* H5T_FLOAT */
+        case H5T_REFERENCE: {
+            /* Convert each element to a list */
+            jboolean bb;
+            jbyte   *barray = NULL;
+            jobject  jobj   = NULL;
+
+            jsize byteArraySize = (jsize)typeSize;
+            if (typeSize != (size_t)byteArraySize)
+                H5_JNI_FATAL_ERROR(ENVONLY, "translate_atomic_rbuf: overflow of byteArraySize");
+
+            if (NULL == (jobj = ENVPTR->NewByteArray(ENVONLY, byteArraySize)))
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+
+            PIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, &bb,
+                           "translate_atomic_rbuf reference: byte array not pinned");
+
+            HDmemcpy(barray, ((jbyte *)raw_buf), typeSize);
+            if (barray)
+                UNPIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, jobj ? 0 : JNI_ABORT);
+
+            if (jobj) {
+                ENVPTR->CallBooleanMethod(ENVONLY, ret_buf, arrAddMethod, jobj);
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
+                ENVPTR->DeleteLocalRef(ENVONLY, jobj);
+            }
+            break;
+        } /* H5T_REFERENCE */
+        case H5T_STRING: {
+            jobject jobj        = NULL;
+            char   *cstr        = raw_buf;
+            htri_t  is_variable = 0;
+
+            if ((is_variable = H5Tis_variable_str(mem_type_id)) < 0)
+                H5_LIBRARY_ERROR(ENVONLY);
+
+            /* Convert each element */
+            if (is_variable) {
+                char **new_buf = (char *)raw_buf;
```

Comment:
This line looks suspicious to me. Did you mean `char **new_buf = &(char *)raw_buf` or even `char **new_buf = &cstr`? (cstr looks to be unused).

Reply:
yes cstr is now removed.
The assignment is because I couldn't get the compiler to agree and this worked (debugged the actual values of the pointer to verify).

Reply:
The cast here is just wrong since you're assigning a `char *` to a `char **`. It may be that a `char **` is getting passed in for `raw_buf` and making things work by coincidence but this is just not quite right.

Reply:
variable len strings are just weird! The issue is that var string data are pointers , everything else is the data. I haven't tried removing the right hand cast - that might make it less wrong looking.

Reply:
okay I can do that

Reply:
done.